### ISSUE TITLE
Resolve differences between Kona and k3.2

### ISSUE THREAD

**Issue by tavmem** [kevinlawler/kona#226]
Create a compiler flag that enables or disables strict compatibility of kona to k3.2 (see issue 225).

**Comment by tavmem** [comment]
I am changing the title of this issue from
"Compiler flag for kona deviations from k3.2" to
"Resolve difference between Kona and k3.2"
At a minimum, Kona should be able to run all the sample code in www.nsl.com.

As a first step: resolve differences in _vs()
Kevin and I have discussed this and we agree that _vs in Kona should work like _vs in k3.2
The "new and improved" _vs that currently exists in Kona will be remamed to "_vs2" or "_vs1".
This will probably require some editing of sample code or golf solutions in Kona that use the Kona version of _vs.

**Comment by tavmem** [comment]
Well, first of all, neither "_vs2" nor "_vs1" works as a name since Kona requires that all reserved verbs be completely alphbetic (probably to protect "floor").  So, we will rename Kona's "extended" version "_vsx".

There seem to be at least 4 incompatibilities between "x _vs y" in k and "x _vsx y" in Kona.

CASE 1) When x is an integer atom, and y is 0.
in k2.8 and k3.2
```
  2 _vs 0
!0
```
in Kona
```
  2 _vsx 0
,0
```
In the first case, the A+ language agrees more closely with Kona
```
    $mode ascii
     2 unpack  0
 0
```

CASE 2) When x is an atom integer, and y is a vector of non-negative integers.
in k2.8 and k3.2
```
  10 _vs 1 12 123 0
(0 0 1 0
 0 1 2 0
 1 2 3 0)
```
in Kona
```
  10 _vsx 1 12 123 0
(,1
 1 2
 1 2 3
 ,0)
```
in A+
```
     $mode ascii
     10 10 10 10 unpack  1 12 123 0
 0 0 0 0
 0 0 1 0
 0 1 2 0
 1 2 3 0
```

CASE 3) When y is a vector of integers where one (or more) are negative
in k2.8 and k3.2
```
  10 _vs 1 -12 123 0
(0 9 1 0
 0 8 2 0
 1 8 3 0)
```
In Kona
```
  10 _vsx 1 -12 123 0
(,1
 -1 -2
 1 2 3
 ,0)
```
BTW: the behavior of k2.8 and k3.2 (in the third case), mimics the behavior of A+.
```
     $mode ascii
     10 10 10 10 unpack  1, (-12), 123 0
 0 9 0 0
 0 9 1 0
 0 8 2 0
 1 8 3 0
```

CASE 4)  Where y is a negative atom integer.  Here, none of the 3 languages agree.
in k2.8 and k3.2
```
  10 _vs -1
4 2 9 4 9 6 7 2 9 5
  10 _vs -2
4 2 9 4 9 6 7 2 9 4
  10 _vs -2147483646
2 1 4 7 4 8 3 6 5 0
  10 _vs -2147483647
2 1 4 7 4 8 3 6 4 9
  10 _vs -2147483648
2 1 4 7 4 8 3 6 4 9
```
In Kona
```
  10 _vsx -1
,-1
  10 _vsx -2
,-2
  10 _vsx -2147483646
-2 -1 -4 -7 -4 -8 -3 -6 -4 -6
  10 _vsx -2147483647
-2 -1 -4 -7 -4 -8 -3 -6 -4 -7
  10 _vsx -2147483648
-2 -1 -4 -7 -4 -8 -3 -6 -4 -7
```
in A+
```
     $mode ascii
     10 10 10 10 10 10 10 10 10 10 unpack -1
 9 9 9 9 9 9 9 9 9 9
     10 10 10 10 10 10 10 10 10 10 unpack -2
 9 9 9 9 9 9 9 9 9 8
     10 10 10 10 10 10 10 10 10 10 unpack -2147483646
 2 1 4 7 4 8 3 6 4 4
     10 10 10 10 10 10 10 10 10 10 unpack -2147483647
 2 1 4 7 4 8 3 6 4 3
     10 10 10 10 10 10 10 10 10 10 unpack -2147483648
 2 1 4 7 4 8 3 6 4 2
```

Next step:  Set up the 2 versions in Kona ( _vs and _vsx ) where, initally, only the Case-1 incompatibility for _vs is resolved.  Then work on cases 2, 3, and 4, and on changing any existing references of _vs in any sample Kona code or golf solutions to _vsx for continuity.

**Comment by tavmem** [comment]
See commit 9c976a48fa0524571bfae1e5ee62d997e02a4f9c

**Comment by tavmem** [comment]
Only found 1 instance of _vs used in Project Euler Golf (problem 36).
Has been changed to _vsx
Did not find any other reference to _vs (in documentation or sample code) in the Wiki.
If someone does find any, please change it, or alert me, and I will change it.

**Comment by bakul** [comment]
As usual I agree with Arthur Whitney's choices!

Case 1: k shows the *minimal* vector. kona doesn't. 
Case 2: vectors are expanded to the same length by appending as many zeroes as needed but at least one vector doesn't start with a 0.
Case 4: the operation we want is modulus, not remainder. Result of mod n is a number in range [0...n).
Case 3 is fixed once the other cases are fixed.

I think kona's _vs should be retired.

**Comment by bakul** [comment]
Given b, some base and v, a vector of  natural numbers, in K the following yields the original v.

```
b _sv b _vs v
```
In Kona _sv behaves as in K , but _vs doesn't so the above expression fails. This is further evidence that Kona's original _vs design was not quite right.

**Comment by tavmem** [comment]
Yes, I agree that the original implementation of _vs in Kona (now renamed _vsx) raises issues.  The primary one (in my opinion) is that code written for k3 or k2 does not run on Kona.  The current objective: any code in www.nsl.com that runs on k2 or k3 should run on Kona w/o modification.  During this transition, we will attempt to provide some continuity for any current users of Kona, and for Kona solutions to Euler golf code.

Once this is done, we can reconsider retiring Kona's _vsx.  (It's probably a good idea.)

BTW: In k2.8 and k3.2, _sv is not the complete inverse of _vs (when used with negative integers):
```
  10 _sv 10 _vs 1 12 123 0
1 12 123 0
  10 _sv 10 _vs 1 -12 123 0
1 988 123 0
```
But, as you pointed out, Kona doesn't work even when all the integers are non-negative.
```
  10 _sv 10 _vsx 1 12 123 0
length error
{{z+y*x}/[0;x;y]}
   ^
>  
```

**Comment by tavmem** [comment]
Of historical interest:
The handling of the negative integer case is a direct carry-over from A+
```
     $mode ascii
     10 10 10 unpack  1,(-12), 123 0
 0 9 1 0
 0 8 2 0
 1 8 3 0
     10 pack 10 10 10 unpack  1,(-12), 123 0
 1 988 123 0
```

**Comment by tavmem** [comment]
Also of historical interest:
The implementation of 3 different modes for providing input to A+ (apl, ascii, unicode) might provide some insight into the implementation of q in k.
How does it work in A+ ?
Are there alternalte parsers, or is some other mechanism used?
